### Romberg method

See in *PDF* folder for precise symbolic expressions

In [1]:
import numpy as np
from scipy.integrate import quad

In [2]:
def integral(f, a, b, h): # Order 2 integration method

    ba = b - a
    N = round(ba / h)
    h = ba / N # Recalculate h to match the exact grid
    x = np.linspace(a, b, N, endpoint = False)
    midpoints = x + h / 2.0

    I = h * np.sum(f(midpoints))

    return I

In [3]:
f = lambda x: np.sqrt(x * np.tan(x)) # f(x) = sqrt(x * tan(x))
a, b = 0.0, np.pi / 3.0 # a = 0, b = pi / 3

h = 1e-02 # h = 0.01, not so small
N = 5
Ih = []
for _ in range(N + 1): # Computing I(h), I(h/2), I(h/4), ...
    x = np.arange(a, b + h, h)
    Ih.append(
        integral(
            f = f,
            a = a,
            b = b,
            h = h
        )
    )

    h /= 2.0

Ih

[np.float64(0.6148719351566976),
 np.float64(0.6148756472806158),
 np.float64(0.6148765887600253),
 np.float64(0.6148768226390339),
 np.float64(0.6148768811089709),
 np.float64(0.6148768957235581)]

In [4]:
I_romb1 = (- Ih[0] + 4.0 * Ih[1]) / 3.0 # One extrapolation
I_romb1, abs(I_romb1 - Ih[0])

(np.float64(0.6148768846552553), np.float64(4.949498557671461e-06))

In [5]:
I_romb2 = (Ih[0] - 20.0 * Ih[1] + 64.0 * Ih[2]) / 45.0 # Two extrapolations
I_romb2, abs(I_romb2 - I_romb1), abs(I_romb2 - Ih[0])

(np.float64(0.6148769037819111),
 np.float64(1.9126655814716287e-08),
 np.float64(4.9686252134861775e-06))

In [6]:
I_romb3 = (- Ih[0] + 84.0 * Ih[1] - 1344.0 * Ih[2] + 4096.0 * Ih[3]) / 2835.0 # Three extrapolations
I_romb3, abs(I_romb3 - I_romb2), abs(I_romb3 - Ih[0])

(np.float64(0.6148769004135534),
 np.float64(3.3683577038701173e-09),
 np.float64(4.965256855782307e-06))

In [7]:
I_romb4 = (Ih[0] - 340.0 * Ih[1] + 22848.0 * Ih[2] - 348160.0 * Ih[3] + 1048576.0 * Ih[4]) / 722925.0 # Four extrapolations
I_romb4, abs(I_romb4 - I_romb3), abs(I_romb4 - Ih[0])

(np.float64(0.6148769006018093),
 np.float64(1.8825596637128683e-10),
 np.float64(4.965445111748679e-06))

In [8]:
real, err = quad(f, a, b) # Precise value
real, err

(0.6148769005989663, 6.826504923551261e-15)

In [ ]:
# Romb4 is much better
abs(I_romb4 - real), abs(Ih[0] - real)

(np.float64(2.843059121460101e-12), np.float64(4.965442268689557e-06))

Best Romberg result computed:

$$
    \int_0^{\frac{\pi}{3}} \, \sqrt{x \, \tan{x}} \, dx = 0.6148769006 \pm 0.0000000002
$$